api/v1alpha1/release_types.go

```diff
@@ -138,128 +144,238 @@ type Release struct {
 	Status ReleaseStatus `json:"status,omitempty"`
 }
 
-// HasStarted checks whether the Release has a valid start time set in its status.
-func (r *Release) HasStarted() bool {
-	return r.Status.StartTime != nil && !r.Status.StartTime.IsZero()
+// HasDeploymentFinished checks whether the Release deployment has finished.
+func (r *Release) HasDeploymentFinished() bool {
+	return r.hasPhaseFinished(deployedConditionType)
+}
+
+// HasEveryPostActionExecutionFinished checks whether the Release post-actions execution has finished.
+func (r *Release) HasEveryPostActionExecutionFinished() bool {
+	return r.hasPhaseFinished(postActionsExecutedConditionType)
+}
+
+// HasProcessingFinished checks whether the Release processing has finished.
+func (r *Release) HasProcessingFinished() bool {
+	return r.hasPhaseFinished(processedConditionType)
 }
 
-// HasSucceeded checks whether the Release has succeeded or not.
-func (r *Release) HasSucceeded() bool {
-	return meta.IsStatusConditionTrue(r.Status.Conditions, releaseConditionType)
+// HasReleaseFinished checks whether the Release has finished.
+func (r *Release) HasReleaseFinished() bool {
+	return r.hasPhaseFinished(releasedConditionType)
 }
 
-// IsDeployed checks whether the Release has been successfully deployed via GitOps.
+// HasValidationFinished checks whether the Release validation has finished.
+func (r *Release) HasValidationFinished() bool {
+	return r.hasPhaseFinished(validatedConditionType)
+}
+
+// IsDeployed checks whether the Release deployment has finished successfully.
 func (r *Release) IsDeployed() bool {
-	condition := meta.FindStatusCondition(r.Status.Conditions, applicationapiv1alpha1.ComponentDeploymentConditionAllComponentsDeployed)
-	return condition != nil && condition.Status == metav1.ConditionTrue
+	return meta.IsStatusConditionTrue(r.Status.Conditions, deployedConditionType.String())
 }
 
-// IsDeploying checks whether the Release has a valid start time for the deployment set in its status.
+// IsDeploying checks whether the Release deployment is progressing.
 func (r *Release) IsDeploying() bool {
-	return r.Status.DeploymentStartTime != nil && !r.Status.DeploymentStartTime.IsZero()
+	return r.isPhaseProgressing(deployedConditionType)
+}
+
+// IsEveryPostActionExecuted checks whether the Release post-actions execution has finished successfully.
+func (r *Release) IsEveryPostActionExecuted() bool {
+	return meta.IsStatusConditionTrue(r.Status.Conditions, postActionsExecutedConditionType.String())
+}
+
+// IsEachPostActionExecuting checks whether the Release post-actions execution is progressing.
+func (r *Release) IsEachPostActionExecuting() bool {
```

Comment:
PostActions will be basically a single Pipeline with multiple actions (tasks) so it's difficult to name the function without using `are`. That's why I called it `isEach` before if we are executing a pipeline, each of the actions (tasks) will be executing. From your options I'd go with `IsAnyPostActionExecuting` but I think it's more confusing tbh.

Reply:
If the "post actions" are contained in a single pipelinerun, why can't we call it "post action" .. the operator is doing one action - starting a single pipelinerun

Reply:
Cause they could be unrelated actions (eg. posting to slack and pushing something to some other place).

Reply:
then by that logic processing should be processings because they could be doing unrelated actions too. I think it makes things nice to use a singular word because then all the functions can be nicely and similarly named, and if it is just one pipelinerun, we can get away with using the singular, even if multiple tasks are in it